### PR TITLE
Add extensible chart markings support

### PR DESCRIPTION
**Pull request by rfinnie**
* [x] (_Not a large audience, but relevant to plugin development_)Your changes are not possible to do through a plugin and relevant
    to a large audience (ideally all users of OctoPrint)
  * [x] If your changes are large or otherwise disruptive: You have
    made sure your changes don't interfere with current development by
    talking it through with the maintainers, e.g. through a
    Brainstorming ticket
  * [x] Your PR targets OctoPrint's `devel` branch if it's a completely
    new feature, or `maintenance` if it's a bug fix or improvement of
    existing functionality for the current stable version (no PRs
    against `master` or anything else please)
  * [x] Your PR was opened from a custom branch on your repository
    (no PRs from your version of `master`, `maintenance`, or `devel`
    please), e.g. `dev/my_new_feature` or `fix/my_bugfix`
  * [x] Your PR only contains relevant changes: no unrelated files,
    no dead code, ideally only one commit - rebase and squash your PR
    if necessary!
  * [x] Your changes follow the existing coding style
  * [x] If your changes include style sheets: You have modified the
    `.less` source files, not the `.css` files (those are generated
    with `lessc`)
  * [x] You have tested your changes (please state how!) - ideally you
    have added unit tests
  * [x] You have run the existing unit tests against your changes and
    nothing broke
  * [x] (_Already in there_)You have added yourself to the `AUTHORS.md` file :)

#### What does this PR do and why is it necessary?

Add extensible chart markings support
    
- Replace add_marking with CHART_MARKED event
- Allow for arbitrary markings in the JS UI

Note that this removes `add_marking()`.  In #4657 I said it would be easiest to just extend it, but since plugin maintainers will need to have a check for backwards compatibility anyway, extending it would require them to catch `TypeError` for missing kwargs, which IMHO was too broad.  As it was impossible to use `add_marking()` for plugins in the past, nothing could have used it and it's safe to replace with an event.

~~The only thing I'm not totally happy with is moving style functionality (the colors) into functional code, but I think the additional functionality to plugin writers is worth it.~~

Edit: moved the colors to CSS, with a default color of grey, and removed all color mentions from the code path.  If plugin authors want, they can define custom CSS for `.temperature-mark-type-{type}`

#### How was it tested? How can it be tested by the reviewer?

- Verified identical visual functionality for all existing built-in markings (print, done, cancel, pause, resume).
- Simple plugin:

```python
from octoprint.events import Events
import octoprint.plugin
import octoprint.util


class TestMarkings(octoprint.plugin.StartupPlugin):
    def on_after_startup(self):
        if hasattr(Events, "CHART_MARKED"):
            timer = octoprint.util.RepeatedTimer(30, self.example_recurring)
            timer.start()

    def example_recurring(self):
        self._event_bus.fire(
            Events.CHART_MARKED,
            {
                "type": "testmarkings_example",
                "label": "Example",
            },
        )


__plugin_name__ = "TestMarkings"
__plugin_pythoncompat__ = ">=3,<4"


def __plugin_load__():
    global __plugin_implementation__
    __plugin_implementation__ = TestMarkings()
```

#### Any background context you want to provide?

Per original feature request:

> I maintain [a plugin](https://plugins.octoprint.org/plugins/bedcooldown/) which begins the bed cooldown shortly before the end of the print, and it would be a good addition to show on the chart when that cooldown begins.

#### What are the relevant tickets if any?

- #4657 (My feature request)
- #4382 (Original PR for chart markings)

#### Screenshots (if appropriate)

image

#### Further notes

I'm not sure how "versionadded" is added for documentation updates; I just put "1.9.0" for the time being.

### CODE REVIEW

**Comment by rfinnie** [comment]
Sorry, rebased from devel to maintenance.  I don't know why I was thinking of this as being completely new functionality :)

**Comment by cp2004** [comment]
One issue with switching the adding of the labels from the JS viewmodel to the server-side events system, is that the `gettext` translations won't work anymore. Because there is no language context there, they will remain as their English defaults. Currently, they are translated by the frontend. `gettext` only functions correctly when called as part of a flask request/with flask context.

**Comment by rfinnie** [comment]
Hmm, is that correct? I've switched over the language to German and the server-side `flask_babel.gettext()` appears to correctly translate the markings:

image

**Comment by cp2004** [comment]
I'll have to double check then - perhaps it varies based on whether you set the server language vs. user language vs. requesting in the URL with `?i18n=` (may not be exactly that - I can't remember...!).

**Comment by rfinnie** [comment]
Cool, thanks, it's entirely possible there's an edge case I haven't encountered yet (at least i18n-wise).

**Comment by cp2004** [comment]
So I've just had a look at this for a test, noticed a couple of things:

* The colours of the lines don't seem to be working correctly. In your screenshot in https://github.com/OctoPrint/OctoPrint/pull/4658#issuecomment-1267289386 they are grey as well - the labels are working OK, but the lines aren't.
* My comment about the translations was mostly correct. Translating server side, outside the request context means that it only respects the default language of the server, not an individual user's language or with the request values `l10n=<language code>`. Now this is not the only place where this is the case in OctoPrint, there are certain things that are already not respecting the translations the same way (such as the permissions roles). I'll leave it up to @foosel to decide which is best in this case.

**Comment by rfinnie** [comment]
Thanks for looking into this.

Re: language, agreed, we should defer to @foosel on it.  I think having the labels server-side is the right way to do it since the mechanism for core and plugins becomes the same, but I don't know how critical HTTP Language / query option overrides should be considered vs OctoPrint-wide language.

Re: lines, turns out that's because there's no way to apply a style to canvas stuff, and while the flag is an HTML element, turns out the line is part of the canvas.  I hadn't even noticed it until you pointed it out, and while it is a visual regression, IMHO it's acceptable given the functionality trade-off.

**Comment by jneilliii** [comment]
>Re: lines, turns out that's because there's no way to apply a style to canvas stuff, and while the flag is an HTML element, turns out the line is part of the canvas. I hadn't even noticed it until you pointed it out, and while it is a visual regression, IMHO it's acceptable given the functionality trade-off.

How did the previous implementation do it? Seems it's able to color the lines fine, so maybe just needs a little more tweaking.

image

File: src/octoprint/static/js/app/viewmodels/temperature.js [hunk 1]
```diff
@@ -520,7 +497,7 @@ $(function () {
                 }
 
                 return {
-                    color: self.marking_props[mark.type].color,
+                    color: "#aeb6bf",
```
Comment (jneilliii) [inline src/octoprint/static/js/app/viewmodels/temperature.js:500]: I think this is where you need adjust the code to have the line the same color? You're forcing the gray.

Comment (rfinnie) [reply src/octoprint/static/js/app/viewmodels/temperature.js]: tl;dr this ended with me going "oh right, you can extract CSS properties via JS!" :)  Fixed to match the corresponding CSS.

image

**Comment by foosel** [comment]
> Re: language, agreed, we should defer to @foosel on it. I think having the labels server-side is the right way to do it since the mechanism for core and plugins becomes the same, but I don't know how critical HTTP Language / query option overrides should be considered vs OctoPrint-wide language.

I was absolutely sure I had already replied to this, but maybe the response just sat unsent in some tab and then went the way of the Dodo on a browser restart.

Anyhow... the way I would go about this is using the same approach I use for the printer state translations. Dummy translations:

https://github.com/OctoPrint/OctoPrint/blob/a68916f05a251052e139486df9932ecb246ae591/src/octoprint/static/js/app/main.js#L257-L281

So, have that somewhere on the frontend, send language agnostic (aka english) strings from the backend (better for an API anyhow) and then use gettext on the dynamic string on the frontend, with a working language context.

**Comment by rfinnie** [comment]
> So, have that somewhere on the frontend, send language agnostic (aka english) strings from the backend (better for an API anyhow) and then use gettext on the dynamic string on the frontend, with a working language context.

Thanks @foosel, PR updated, and confirmed mark labels work with either server-wide language or `?l10n=de`.

**Review by foosel — APPROVED** [review]

**Comment by foosel** [comment]
Looking good, merged, thank you very much!

**Comment by foosel** [comment]
Currently "something else" in this case is "Gina will do it when she looks into merging and releasing stuff" 😅

**Comment by cp2004** [comment]
As long as you either update the less, or update the less *and* css, it's alright. Just not the css on it's own.

**Comment by rfinnie** [comment]
@foosel If you are interested in highlighting this functionality with the 1.9.0 release, I just wanted to let you know I've released OctoPrint-BedCooldown 0.6.0 today, which has support for it, and (IMHO) best practices:

- Conditional support (does not break pre-1.9.0 installations)
- Stylized (sets its own marking color)
- Language support (no actual translations, but the gettext framework is in place)

image

**Comment by foosel** [comment]
@rfinnie That's awesome! Would it be cool to use this screenshot in the release announcement?

**Comment by rfinnie** [comment]
@foosel Yep, no problem to include that screenshot.
